File: swift/day14_part2_2018.swift
import Foundation

let fileURL = URL(fileURLWithPath: "input.txt")
let input = try String(contentsOf: fileURL).trimmingCharacters(in: .whitespacesAndNewlines)

var scoreboard = [3, 7]
var elf1 = 0
var elf2 = 1
let inputLen = input.count
var inputSequence = [Int]()

for i in 0..<inputLen {
    inputSequence.append(Int(String(input[input.index(input.startIndex, offsetBy: i)]))!)
}

func checkSequence(scoreboard: [Int], sequence: [Int]) -> Bool {
    if scoreboard.count < sequence.count {
        return false
    }
    let start = scoreboard.count - sequence.count
    for i in 0..<sequence.count {
        if scoreboard[start + i] != sequence[i] {
            return false
        }
    }
    return true
}

while true {
    let newScore = scoreboard[elf1] + scoreboard[elf2]
    if newScore >= 10 {
        scoreboard.append(newScore / 10)
        if checkSequence(scoreboard: scoreboard, sequence: inputSequence) {
            break
        }
    }
    scoreboard.append(newScore % 10)
    if checkSequence(scoreboard: scoreboard, sequence: inputSequence) {
        break
    }

    elf1 = (elf1 + scoreboard[elf1] + 1) % scoreboard.count
    elf2 = (elf2 + scoreboard[elf2] + 1) % scoreboard.count
}

print(scoreboard.count - inputLen)

File: ruby/day14_part2_2018.rb
def main
  input = File.read("input.txt").strip
  scoreboard = [3, 7]
  elf1, elf2 = 0, 1
  input_sequence = input.chars.map(&:to_i)

  loop do
    new_score = scoreboard[elf1] + scoreboard[elf2]
    if new_score >= 10
      scoreboard << new_score / 10
      break if check_sequence(scoreboard, input_sequence)
    end
    scoreboard << new_score % 10
    break if check_sequence(scoreboard, input_sequence)

    elf1 = (elf1 + scoreboard[elf1] + 1) % scoreboard.size
    elf2 = (elf2 + scoreboard[elf2] + 1) % scoreboard.size
  end

  puts scoreboard.size - input_sequence.size
end

def check_sequence(scoreboard, sequence)
  return false if scoreboard.size < sequence.size

  start = scoreboard.size - sequence.size
  sequence.each_with_index.all? { |value, index| scoreboard[start + index] == value }
end

main